libraries/chain_kv/include/b1/session/shared_bytes.hpp

```diff
@@ -139,6 +141,39 @@ inline shared_bytes shared_bytes::next() const {
    return eosio::session::shared_bytes(buffer.data(), buffer.size());
 }
 
+inline shared_bytes shared_bytes::next_sub_key() const {
+   auto buffer = std::vector<unsigned char>{ std::begin(*this), std::end(*this) };
+   buffer.push_back('\0');
+
+   return eosio::session::shared_bytes(buffer.data(), buffer.size());
+}
+
+inline shared_bytes shared_bytes::previous() const {
+   if (!*this)
+      return shared_bytes{};
+
+   auto buffer = std::vector<unsigned char>{ std::begin(*this), std::end(*this) };
+
+   int index = buffer.size() - 1;
+   while (true) {
+      auto& val = buffer[index];
+      if (val) {
+         --val;
+         ++index; // ensure we increment any indexes past this one
+         break;
+      }
+      if (--index < 0) {
+         buffer.pop_back();
+         index = buffer.size();
+      }
+   }
+   while (index < buffer.size()) {
+      buffer[index] = std::numeric_limits<unsigned char>::max();
+   }
```

Comment:
I had this wrong.  The logic of next is really incorrect I believe.  It should just increment at the index, and then rollback to 0, and then increment the next lower index.  In the corner case of 0xfff..., that is less straight forward.  I am thinking that in that case we would want it to then just append 0xff at the end.  The new previous() logic is to decrement each index (starting at the last), if it is greater than 0 then exit, otherwise it rolls back to 0xff and then goes to the next lower index and repeats.  For 0x0000...00, it just shortens the vector by 1

Reply:
The implementation of `next` is correct according to the definition I gave above (with the exception of `FF*` for which the "correct" result is undefined) and this matches how it is actually used.  Based on how `previous` is used, (in `chain_plugin.cpp` `read_only::get_table_by_scope`), I don't think there is a way to define `previous` that will work correctly.